File: go/test/deal_field_entity_test.go
package sdktest

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/pipedrive-sdk/go"
	"github.com/voxgig-sdk/pipedrive-sdk/go/core"

	vs "github.com/voxgig-sdk/pipedrive-sdk/go/utility/struct"
)

func TestDealFieldEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.DealField(nil)
		if ent == nil {
			t.Fatal("expected non-nil DealFieldEntity")
		}
	})

	// Feature #4: the entity Stream(action, ...) method runs the op pipeline and
	// returns a channel over result items. With the streaming feature active it
	// yields the feature's incremental output; otherwise it falls back to the
	// materialised list so Stream always yields.
	t.Run("stream", func(t *testing.T) {
		seed := map[string]any{
			"entity": map[string]any{
				"deal_field": map[string]any{
					"s1": map[string]any{"id": "s1"},
					"s2": map[string]any{"id": "s2"},
					"s3": map[string]any{"id": "s3"},
				},
			},
		}

		// Fallback: streaming inactive -> yields the materialised list items.
		base := sdk.TestSDK(seed, nil)
		var seen []any
		for item := range base.DealField(nil).Stream("list", nil, nil) {
			seen = append(seen, item)
		}
		if len(seen) != 3 {
			t.Fatalf("expected 3 streamed items, got %d", len(seen))
		}

		// Inbound: streaming active -> yields each item from the feature iterator.
		hasStreaming := false
		if fm, ok := core.SharedConfig()["feature"].(map[string]any); ok {
			_, hasStreaming = fm["streaming"]
		}
		if hasStreaming {
			streamSdk := sdk.TestSDK(seed, map[string]any{
				"feature": map[string]any{"streaming": map[string]any{"active": true}},
			})
			var got []any
			for item := range streamSdk.DealField(nil).Stream("list", nil, nil) {
				if sub, ok := item.([]any); ok {
					got = append(got, sub...)
				} else {
					got = append(got, item)
				}
			}
			if len(got) != 3 {
				t.Fatalf("expected 3 items via streaming feature, got %d", len(got))
			}
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := deal_fieldBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "list", "update", "load", "remove"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "deal_field." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set PIPEDRIVE_TEST_DEAL_FIELD_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		dealFieldRef01Ent := client.DealField(nil)
		dealFieldRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath(setup.data, []any{"new", "deal_field"}), "deal_field_ref01"))

		dealFieldRef01DataResult, err := dealFieldRef01Ent.Create(dealFieldRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		dealFieldRef01Data = core.ToMapAny(entityData(dealFieldRef01DataResult))
		if dealFieldRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}
		if dealFieldRef01Data["id"] == nil {
			t.Fatal("expected created entity to have an id")
		}

		// LIST
		dealFieldRef01Match := map[string]any{}

		dealFieldRef01ListResult, err := dealFieldRef01Ent.List(dealFieldRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		dealFieldRef01List, dealFieldRef01ListOk := dealFieldRef01ListResult.([]any)
		if !dealFieldRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", dealFieldRef01ListResult)
		}

		foundItem := vs.Select(entityListToData(dealFieldRef01List), map[string]any{"id": dealFieldRef01Data["id"]})
		if vs.IsEmpty(foundItem) {
			t.Fatal("expected to find created entity in list")
		}

		// UPDATE
		dealFieldRef01DataUp0Up := map[string]any{
			"id": dealFieldRef01Data["id"],
		}

		dealFieldRef01MarkdefUp0Name := "name"
		dealFieldRef01MarkdefUp0Value := fmt.Sprintf("Mark01-deal_field_ref01_%d", setup.now)
		dealFieldRef01DataUp0Up[dealFieldRef01MarkdefUp0Name] = dealFieldRef01MarkdefUp0Value

		dealFieldRef01ResdataUp0Result, err := dealFieldRef01Ent.Update(dealFieldRef01DataUp0Up, nil)
		if err != nil {
			t.Fatalf("update failed: %v", err)
		}
		dealFieldRef01ResdataUp0 := core.ToMapAny(entityData(dealFieldRef01ResdataUp0Result))
		if dealFieldRef01ResdataUp0 == nil {
			t.Fatal("expected update result to be a map")
		}
		if dealFieldRef01ResdataUp0["id"] != dealFieldRef01DataUp0Up["id"] {
			t.Fatal("expected update result id to match")
		}
		if dealFieldRef01ResdataUp0[dealFieldRef01MarkdefUp0Name] != dealFieldRef01MarkdefUp0Value {
			t.Fatalf("expected %s to be updated, got %v", dealFieldRef01MarkdefUp0Name, dealFieldRef01ResdataUp0[dealFieldRef01MarkdefUp0Name])
		}

		// LOAD
		dealFieldRef01MatchDt0 := map[string]any{
			"id": dealFieldRef01Data["id"],
		}
		dealFieldRef01DataDt0Loaded, err := dealFieldRef01Ent.Load(dealFieldRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		dealFieldRef01DataDt0LoadResult := core.ToMapAny(entityData(dealFieldRef01DataDt0Loaded))
		if dealFieldRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if dealFieldRef01DataDt0LoadResult["id"] != dealFieldRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

		// REMOVE
		dealFieldRef01MatchRm0 := map[string]any{
			"id": dealFieldRef01Data["id"],
		}
		_, err = dealFieldRef01Ent.Remove(dealFieldRef01MatchRm0, nil)
		if err != nil {
			t.Fatalf("remove failed: %v", err)
		}

		// LIST
		dealFieldRef01MatchRt0 := map[string]any{}

		dealFieldRef01ListRt0Result, err := dealFieldRef01Ent.List(dealFieldRef01MatchRt0, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		dealFieldRef01ListRt0, dealFieldRef01ListRt0Ok := dealFieldRef01ListRt0Result.([]any)
		if !dealFieldRef01ListRt0Ok {
			t.Fatalf("expected list result to be an array, got %T", dealFieldRef01ListRt0Result)
		}

		notFoundItem := vs.Select(entityListToData(dealFieldRef01ListRt0), map[string]any{"id": dealFieldRef01Data["id"]})
		if !vs.IsEmpty(notFoundItem) {
			t.Fatal("expected removed entity to not be in list")
		}

	})
}

func deal_fieldBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "deal_field", "DealFieldTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read deal_field test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse deal_field test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap, _ := vs.Transform(
		[]any{"deal_field01", "deal_field02", "deal_field03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("PIPEDRIVE_TEST_DEAL_FIELD_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"PIPEDRIVE_TEST_DEAL_FIELD_ENTID": idmap,
		"PIPEDRIVE_TEST_LIVE":      "FALSE",
		"PIPEDRIVE_TEST_EXPLAIN":   "FALSE",
		"PIPEDRIVE_APIKEY":         "",
	})

	idmapResolved := core.ToMapAny(env["PIPEDRIVE_TEST_DEAL_FIELD_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["PIPEDRIVE_TEST_LIVE"] == "TRUE" {
		// An empty map, not a nil one: Merge returns nil when its last entry
		// is nil, and BasicSetup is normally called with no extras - so a
		// bare nil silently discarded the apikey and server values below.
		extraOpts := extra
		if extraOpts == nil {
			extraOpts = map[string]any{}
		}

		mergedOpts := vs.Merge([]any{
			// liveClientOptions() FIRST, so the generated fields below win:
			// sdk-test-control.json's test.client.options adds to the live
			// client, it does not redirect it.
			liveClientOptions(),
			map[string]any{
				"apikey": env["PIPEDRIVE_APIKEY"],
			},
			extraOpts,
		})
		client = sdk.NewPipedriveSDK(core.ToMapAny(mergedOpts))
	}

	live := env["PIPEDRIVE_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["PIPEDRIVE_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: php/test/DealFieldEntityTest.php
<?php
declare(strict_types=1);

// DealField entity test

require_once __DIR__ . '/../pipedrive_sdk.php';
require_once __DIR__ . '/Runner.php';

use PHPUnit\Framework\TestCase;
use Voxgig\Struct\Struct as Vs;

class DealFieldEntityTest extends TestCase
{
    public function test_create_instance(): void
    {
        $testsdk = PipedriveSDK::test(null, null);
        $ent = $testsdk->DealField(null);
        $this->assertNotNull($ent);
    }

    // Feature #4: the entity stream(action, ...) method runs the op pipeline
    // and yields result items. With the streaming feature active it yields the
    // feature's incremental output; otherwise it falls back to the materialised
    // list so stream always yields.
    public function test_stream(): void
    {
        $seed = [
            "entity" => [
                "deal_field" => [
                    "s1" => ["id" => "s1"],
                    "s2" => ["id" => "s2"],
                    "s3" => ["id" => "s3"],
                ],
            ],
        ];

        // Fallback: streaming inactive -> yields the materialised list items.
        $base = PipedriveSDK::test($seed, null);
        $seen = iterator_to_array($base->DealField(null)->stream("list", null, null), false);
        $this->assertCount(3, $seen);

        // Inbound: streaming active -> yields each item from the feature.
        $cfg = PipedriveConfig::shared_config();
        if (isset($cfg["feature"]) && is_array($cfg["feature"]) && isset($cfg["feature"]["streaming"])) {
            $sdk = PipedriveSDK::test($seed, ["feature" => ["streaming" => ["active" => true]]]);
            $got = [];
            foreach ($sdk->DealField(null)->stream("list", null, null) as $item) {
                if (is_array($item) && array_is_list($item)) {
                    foreach ($item as $sub) {
                        $got[] = $sub;
                    }
                } else {
                    $got[] = $item;
                }
            }
            $this->assertCount(3, $got);
        }
    }

    public function test_basic_flow(): void
    {
        $setup = deal_field_basic_setup(null);
        // Per-op sdk-test-control.json skip.
        $_live = !empty($setup["live"]);
        foreach (["create", "list", "update", "load", "remove"] as $_op) {
            [$_shouldSkip, $_reason] = Runner::is_control_skipped("entityOp", "deal_field." . $_op, $_live ? "live" : "unit");
            if ($_shouldSkip) {
                $this->markTestSkipped($_reason ?? "skipped via sdk-test-control.json");
                return;
            }
        }
        // The basic flow consumes synthetic IDs from the fixture. In live mode
        // without an *_ENTID env override, those IDs hit the live API and 4xx.
        if (!empty($setup["synthetic_only"])) {
            $this->markTestSkipped("live entity test uses synthetic IDs from fixture — set PIPEDRIVE_TEST_DEAL_FIELD_ENTID JSON to run live");
            return;
        }
        $client = $setup["client"];

        // CREATE
        $deal_field_ref01_ent = $client->DealField(null);
        $deal_field_ref01_data = Helpers::to_map(Vs::getprop(
            Vs::getpath($setup["data"], "new.deal_field"), "deal_field_ref01"));

        $deal_field_ref01_data_result = $deal_field_ref01_ent->create($deal_field_ref01_data, null);
        $deal_field_ref01_data = Helpers::to_map(is_object($deal_field_ref01_data_result) && method_exists($deal_field_ref01_data_result, 'data_get') ? $deal_field_ref01_data_result->data_get() : $deal_field_ref01_data_result);
        $this->assertNotNull($deal_field_ref01_data);
        $this->assertNotNull($deal_field_ref01_data["id"]);

        // LIST
        $deal_field_ref01_match = [];

        $deal_field_ref01_list_result = $deal_field_ref01_ent->list($deal_field_ref01_match, null);
        $this->assertIsArray($deal_field_ref01_list_result);

        $found_item = sdk_select(
            Runner::entity_list_to_data($deal_field_ref01_list_result),
            ["id" => $deal_field_ref01_data["id"]]);
        $this->assertNotEmpty($found_item);

        // UPDATE
        $deal_field_ref01_data_up0_up = [
            "id" => $deal_field_ref01_data["id"],
        ];

        $deal_field_ref01_markdef_up0_name = "name";
        $deal_field_ref01_markdef_up0_value = "Mark01-deal_field_ref01_" . $setup["now"];
        $deal_field_ref01_data_up0_up[$deal_field_ref01_markdef_up0_name] = $deal_field_ref01_markdef_up0_value;

        $deal_field_ref01_resdata_up0_result = $deal_field_ref01_ent->update($deal_field_ref01_data_up0_up, null);
        $deal_field_ref01_resdata_up0 = Helpers::to_map(is_object($deal_field_ref01_resdata_up0_result) && method_exists($deal_field_ref01_resdata_up0_result, 'data_get') ? $deal_field_ref01_resdata_up0_result->data_get() : $deal_field_ref01_resdata_up0_result);
        $this->assertNotNull($deal_field_ref01_resdata_up0);
        $this->assertEquals($deal_field_ref01_resdata_up0["id"], $deal_field_ref01_data_up0_up["id"]);
        $this->assertEquals($deal_field_ref01_resdata_up0[$deal_field_ref01_markdef_up0_name], $deal_field_ref01_markdef_up0_value);

        // LOAD
        $deal_field_ref01_match_dt0 = [
            "id" => $deal_field_ref01_data["id"],
        ];
        $deal_field_ref01_data_dt0_loaded = $deal_field_ref01_ent->load($deal_field_ref01_match_dt0, null);
        $deal_field_ref01_data_dt0_load_result = Helpers::to_map(is_object($deal_field_ref01_data_dt0_loaded) && method_exists($deal_field_ref01_data_dt0_loaded, 'data_get') ? $deal_field_ref01_data_dt0_loaded->data_get() : $deal_field_ref01_data_dt0_loaded);
        $this->assertNotNull($deal_field_ref01_data_dt0_load_result);
        $this->assertEquals($deal_field_ref01_data_dt0_load_result["id"], $deal_field_ref01_data["id"]);

        // REMOVE
        $deal_field_ref01_match_rm0 = [
            "id" => $deal_field_ref01_data["id"],
        ];
        $deal_field_ref01_ent->remove($deal_field_ref01_match_rm0, null);

        // LIST
        $deal_field_ref01_match_rt0 = [];

        $deal_field_ref01_list_rt0_result = $deal_field_ref01_ent->list($deal_field_ref01_match_rt0, null);
        $this->assertIsArray($deal_field_ref01_list_rt0_result);

        $not_found_item = sdk_select(
            Runner::entity_list_to_data($deal_field_ref01_list_rt0_result),
            ["id" => $deal_field_ref01_data["id"]]);
        $this->assertEmpty($not_found_item);

    }
}

function deal_field_basic_setup($extra)
{
    Runner::load_env_local();

    $entity_data_file = __DIR__ . '/../../.sdk/test/entity/deal_field/DealFieldTestData.json';
    $entity_data_source = file_get_contents($entity_data_file);
    $entity_data = json_decode($entity_data_source, true);

    $options = [];
    $options["entity"] = $entity_data["existing"];

    $client = PipedriveSDK::test($options, $extra);

    // Generate idmap.
    $idmap = [];
    foreach (["deal_field01", "deal_field02", "deal_field03"] as $k) {
        $idmap[$k] = strtoupper($k);
    }

    // Detect ENTID env override before envOverride consumes it. When live
    // mode is on without a real override, the basic test runs against synthetic
    // IDs from the fixture and 4xx's. Surface this so the test can skip.
    $entid_env_raw = getenv("PIPEDRIVE_TEST_DEAL_FIELD_ENTID");
    $idmap_overridden = $entid_env_raw !== false && str_starts_with(trim($entid_env_raw), "{");

    $env = Runner::env_override([
        "PIPEDRIVE_TEST_DEAL_FIELD_ENTID" => $idmap,
        "PIPEDRIVE_TEST_LIVE" => "FALSE",
        "PIPEDRIVE_TEST_EXPLAIN" => "FALSE",
        "PIPEDRIVE_APIKEY" => "",
    ]);

    $idmap_resolved = Helpers::to_map(
        $env["PIPEDRIVE_TEST_DEAL_FIELD_ENTID"]);
    if ($idmap_resolved === null) {
        $idmap_resolved = Helpers::to_map($idmap);
    }

    if ($env["PIPEDRIVE_TEST_LIVE"] === "TRUE") {
        $merged_opts = Vs::merge([
            // FIRST, so the generated fields below win: sdk-test-control.json's
            // test.client.options adds to the live client, it does not redirect it.
            Runner::live_client_options(),
            [
                "apikey" => $env["PIPEDRIVE_APIKEY"],
            ],
            // ismap, not a plain "?? []" default: an empty PHP array is a
            // LIST, and a non-map later entry REPLACES the accumulated map in
            // merge - so the no-extras call discarded live_client_options()
            // and the apikey/server map above it.
            Vs::ismap($extra) ? $extra : new \stdClass(),
        ]);
        // "?? []" because merge legitimately answers with a stdClass when every
        // contributing entry is an EMPTY map - an SDK with no apikey and no
        // server variables generates an empty middle entry, so that is the
        // common case, not the edge one. to_map returns null for a non-array by
        // design, and the constructor takes a non-nullable array, so without the
        // fallback every such SDK died on "must be of type array, null given"
        // the moment live mode was switched on. Offline mode never reaches this
        // branch, which is why the offline suite stayed green.
        $client = new PipedriveSDK(Helpers::to_map($merged_opts) ?? []);
    }

    $live = $env["PIPEDRIVE_TEST_LIVE"] === "TRUE";
    return [
        "client" => $client,
        "data" => $entity_data,
        "idmap" => $idmap_resolved,
        "env" => $env,
        "explain" => $env["PIPEDRIVE_TEST_EXPLAIN"] === "TRUE",
        "live" => $live,
        "synthetic_only" => $live && !$idmap_overridden,
        "now" => (int)(microtime(true) * 1000),
    ];
}
